### PR TITLE
Add new safe check to backup/update

### PR DESCRIPTION
**Pull request by fernandog**
Check if show are updating and/or SR is searching for episode before auto-update or manual update runs. Previous checks were only if PP were running and showsupdating (this PR fixes the forced-update check_

### CODE REVIEW

**Comment by SiCKRAGEB0T** [comment]
Thanks for the pull request! Before a real human comes by, please make sure your report has all the below criteria checked

- [x] Give a description on what the PR is for.
- [x] Make sure you send a PR against the DEVELOP branch
- [x] Don't send big changes in one go. Split up big PRs into multiple smaller (easier manageble) PRs

Please make sure you also read [contribution guide](https://github.com/SiCKRAGETV/SickRage/blob/master/contributing.md) and followed all the steps.

Thanks!

**Comment by miigotu** [comment]
So now you want to add logic about show/search queue into versionChecker? Come on man.

**Comment by fernandog** [comment]
Yes. Cant update while shows are being updated or while searching eps.
What is wrong with that?!

**Comment by miigotu** [comment]
So you pause the threads. You are adding a shitload of nonsense in version checker, and there is already stuff that doesn't need to be there. You are only making it worse.

**Comment by fernandog** [comment]
We cant unpause after restart. SR lose  queue

**Comment by miigotu** [comment]
SR loses the queue no matter what, it ALWAYS just closes the thread down to restart anyways.

**Comment by miigotu** [comment]
You pause it, the current queue item finishes, you back up, you unpause queue, restart. its like a normal restart then.

**Comment by fernandog** [comment]
why backup a data that will be updated soon?
what's the problem of wait the thread to finished to backup&update later?

**Comment by fernandog** [comment]
@SiCKRAGETV/moderators please feedback this PR.

File: sickbeard/versionChecker.py [hunk 1]
```diff
@@ -185,12 +185,29 @@ def showupdate_safe(self):
...
+                return False
+            
 
         db_safe = db_safe(self)
         postprocessor_safe = postprocessor_safe(self)
         showupdate_safe = showupdate_safe(self)
+        showqueue_empty = showqueue_empty(self)
 
-        if db_safe == True and postprocessor_safe == True and showupdate_safe == True:
+        if db_safe == True and postprocessor_safe == True and showupdate_safe == True and showqueue_empty == True:
```
Comment (MGaetan89) [inline sickbeard/versionChecker.py]: Simply `and showqueue_empty`?

File: sickbeard/webserve.py [hunk 2]
```diff
@@ -1115,6 +1115,7 @@ def update(self, pid=None):
                                             "Update wasn't successful, not restarting. Check your log for more information.")
         else:
             ui.notifications.message('Not safe to update now. Some threads are running. Check Server Status page')
+            logger.log(u"Not safe to update now. Some threads are running. Check Server Status page")
```
Comment (MGaetan89) [inline sickbeard/webserve.py]: `logger.DEBUG`?

File: sickbeard/versionChecker.py [hunk 3]
```diff
@@ -185,12 +185,29 @@ def showupdate_safe(self):
...
+        def showqueue_empty(self):
+            queuelength = sickbeard.searchQueueScheduler.action.queue_length()
+            failed = int(queuelength['failed'])
+            manual = int(queuelength['manual'])
+            daily = int(queuelength['daily'])
+            backlog = int(queuelength['backlog'])
+            action_size = len(sickbeard.showQueueScheduler.action.queue)
+            queue = failed + manual + daily + backlog + action_size
+            if  queue == 0:
+                logger.log(u"We can proceed with the update. Show queue is empty", logger.DEBUG)
```
Comment (MGaetan89) [inline sickbeard/versionChecker.py]: Do we need to log something if the update can be performed? Don't we already log something when the update starts?

**Comment by MGaetan89** [comment]
Commented the changed before reading the discussion :P
I agree with @miigotu that the `versionChecker` class should stick with its original purpose: check for new version.
Is it possible to send a message/notification to all the running threads to tell them that SR is about to restart/update and wait for them to be stopped?

**Comment by miigotu** [comment]
because the queue can take hours. pause queue, wait current item finishes, backup, update, unpause, restart.

I can't poosibly expect you to understand this ofc. you have no concept of threading or OOP and can't even understand a two part true/false if statement. I'm not gonna argue anymore.

Tired of staying up 10 hours past the time I should go to sleep just to make sure you don't merge some stupid shit.

Go ahead and merge it so I can strip all this bullshit out when I wake up.

**Comment by fernandog** [comment]
@MGaetan89 versionchecker needs to be re-written then. And I don't now how to send this message/notification and wait threads to finished. 

The important here we can't backup&update&restart if threads are running, specially PP and showUpdate

**Comment by MGaetan89** [comment]
@fernandog I don't know either. Not enough Python knowledge for that. The idea would be to notify the others threads that we want to stop and that they HAVE to stop as soon as possible.

**Comment by miigotu** [comment]
@MGaetan89 everything is possible if we just stop letting it be done the wrong way. PR like these don't help, that just make more bugs and more work to do.

**Comment by MGaetan89** [comment]
@miigotu I understand that. I'm just trying to find/suggest an alternative solution that can fix the issue without needing to wait hours.

**Comment by SiCKRAGEB0T** [comment]
Then why not empty all the non-running queue items from the queue, and pause the threads when there are no more queues items in them, and proceed with backup when all threads are paused?

-----Original Message-----

**Comment by miigotu** [comment]
you just tell them to stop and they will, or just pause them all and they will finish their current task. then you use a threading lock to block all the threads are done their current item.

then you backup the files. nothing will be using them because the threads have paused.

then you update.

then you unpause so they can shutdown gracefully just like on a normal restart.

then you restart.

Its like 200 less lines and 200x more effective because its the PROPER WAY

**Comment by fernandog** [comment]
Not enough Python knowledge too
well, tell abeloin we are doing wrong way because I come up with the idea and he help code it. he develop the structure, how it would work and I code it. he is the experienced one. not me.
most users here are here to learn, myself included. My python code was 1-2years ago. Not a real programmer. Sorry I'm not a experienced programmer

**Comment by miigotu** [comment]
@MGaetan89 there isn't even an issue to begin with.

**Comment by fernandog** [comment]
yes, there is. can't backup a file while is locked. the issue is open and you know. wait, you closed

**Comment by miigotu** [comment]
If you don't have enough python knowledge, and aren't experienced, why are you touching code that you obviously don't understand?

And secondly, why do I always need to flip the fuck out before you stop arguing with me when I say something is not right?

I could give a fuck if you are here to learn. That does not make it OK to break shit or make problems worse.

**Comment by fernandog** [comment]
LOL! code is working. didn't break anything. kkkkkk
so fucking funny all this

**Comment by miigotu** [comment]
if the queues are PAUSED, and their current item has FINISHED ---- NOTHING IS LOCKING THE FILES THEN.

**Comment by miigotu** [comment]
so fucking funny that you don't know what the fuck you are doing but you think you know everything.

**Comment by WebSpider** [comment]
Well, versionchecker needs to be redone anyways to get rid of all these popen() calls. 

@fernandog how urgent is this issue, many users impacted with nonfunctional installs?

**Comment by miigotu** [comment]
but oh, wait, let me stay up another 5 hours to babysit and argue

**Comment by miigotu** [comment]
@WebSpider it has never been reported except by him.

**Comment by fernandog** [comment]
low priority @WebSpider 

if I don't know what I'm doing how abeloin approved this?  it's working since was merged. I only added a NEW check.

Feel free to say out of discussion then.

**Comment by fernandog** [comment]
@WebSpider force full update of some shows and hit 'update" let's see what happens.....lol

File: sickbeard/versionChecker.py [hunk 4]
```diff
@@ -185,12 +185,29 @@ def showupdate_safe(self):
             else:
                 logger.log(u"We can't proceed with the update. Shows are being updated", logger.DEBUG)
                 return False
+
```
Comment (WebSpider) [inline sickbeard/versionChecker.py]: Just counting queue items doesn't give a reliable result. You need to close the inflow of new queue items, or at least stop the processing of new queue items. Else, a daily search could start a.microsecond after you decided it is safe to upgrade

**Comment by miigotu** [comment]
which you caused by your last two PR to version checker. just stick to your typos and log lines.

**Comment by fernandog** [comment]
in your opinion. feel free to reproduce without my last PR

**Comment by miigotu** [comment]
if you think by keeping arguing that you are making yourself look better or that you are right then you are more stupid than I thought.

**Comment by WebSpider** [comment]
@fernandog @miigotu I think we all want what's best for end users: Stable product with cool features. 

@fernandog on the review request:I

 don't think this pr solves the problem at hand, which is that we wait for queues to empty, while there is no real reason to queues don't need to be empty, they need to finish the task at hand and not process new stuff, so we can get consistent to disk.

I think it would be better to solve that problem completely by making sure we cannot process items while we update, instead of counting when a queue is empty.

**Comment by fernandog** [comment]
ok. thanks for the input

**Comment by miigotu** [comment]
@WebSpider thanks. this PR would almost never have allowed update, because there is almost always something in one queue or another.

**Comment by WebSpider** [comment]
@fernandog only on my win10 install. Syno works like a charm, so I guess it's a filesystem thing.

**Comment by fernandog** [comment]
@WebSpider mine is linux.

**Pull request closed without merging**